File: 72-edit-distance/72-edit-distance.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        global final
        final = 10**10
        record = dict()
        if word2 == '':
            return len(word1)
        if word1 == '':
            return len(word2)
        def find(word_1, word_2):
            if (word_1, word_2) in record:
                return record[(word_1,word_2)]
            if word_1 == word_2:
                record[(word_1,word_2)] = 0
                return 0
            if not word_2:
                record[(word_1,word_2)] = len(word_1)
                return len(word_1)
            if not word_1:
                record[(word_1,word_2)] = len(word_2)
                return len(word_2)
            if word_1[0] == word_2[0]:
                ans = find(word_1[1:],word_2[1:])
            else:
                #insert
                a = find(word_2[0]+word_1,word_2) + 1
                #delete
                b = find(word_1[1:],word_2) + 1
                #replace
                c = find(word_2[0]+word_1[1:],word_2)+1
                ans = min([a,b,c])
            record[(word_1,word_2)] = ans
            return ans
        return find(word1,word2)
                
                
            
        
                
            

File: 72-edit-distance/72-edit-distance.java
class Solution {
    Map<String, Integer> memo = new HashMap<>();
    int MAX = 20000;
    public int minDistance(String word1, String word2) {
        int startLen = word1.length();
        int targetLen = word2.length();
        String key = word1+"#" + word2;
        
        if(memo.containsKey(key)){
            return memo.get(key);
        }
        
        if(startLen == 0){
            if(targetLen == 0)return 0;
            return targetLen;
        }
        
        if(targetLen == 0)return startLen - targetLen;
        int ans = MAX;
        //three possible cases
        char startFirst = word1.charAt(0), targetFirst = word2.charAt(0);
        if(startFirst == targetFirst){
            ans = Math.min(minDistance(word1.substring(1), word2.substring(1)), ans);
        }else{
            //replace
            ans = Math.min(minDistance(word1.substring(1), word2.substring(1)) + 1, ans);
            
            //delete
            ans = Math.min(minDistance(word1.substring(1), word2) + 1, ans);
            //insert
            ans = Math.min(minDistance(word1, word2.substring(1)) + 1, ans);                  
        }
        memo.put(key, ans);
        return ans;        
    }
}